watchtower-plugin/src/retrier.rs

```diff
@@ -226,22 +389,53 @@ impl Retrier {
                     // unsuccessfully or by manually raising such an error (like when facing a tower misbehavior).
                     log::warn!("Retry strategy gave up for {}. {}", self.tower_id, e);
 
-                    // Retrier failed and should be given up on. Avoid setting the tower status until the retrier is
-                    // deleted/dropped. This way users performing manual retry will get an error as the tower will be
-                    // temporary unreachable.
-                    // We don't need to set the tower status now. Any new appointments we receive will not be retried anyways.
-                    self.set_status(RetrierStatus::Failed);
+                    if e.is_permanent() {
+                        self.set_status(RetrierStatus::Failed);
+                    } else {
+                        log::debug!("Starting to idle");
+                        self.set_status(RetrierStatus::Idle(Instant::now()));
+                        // Clear all pending appointments so they do not waste any memory while idling
+                        self.pending_appointments.lock().unwrap().clear();
+                        self.wt_client
+                            .lock()
+                            .unwrap()
+                            .set_tower_status(self.tower_id, crate::TowerStatus::Unreachable);
+                    }
+
+                    match e {
+                        RetryError::Subscription(_, _) => {
+                            log::info!("Setting {} status as subscription error", self.tower_id);
+                            self.wt_client
+                                .lock()
+                                .unwrap()
+                                .set_tower_status(self.tower_id, TowerStatus::SubscriptionError)
+                        }
+                        RetryError::Misbehaving(p) => {
+                            log::warn!("Cannot recover known tower_id from the appointment receipt. Flagging tower as misbehaving");
+                            self.wt_client
+                                .lock()
+                                .unwrap()
+                                .flag_misbehaving_tower(self.tower_id, p);
+                        }
+                        RetryError::Abandoned => {
+                            log::info!("Skipping retrying abandoned tower {}", self.tower_id)
+                        }
+                        // Notice that setting the state as Unreachable, when needed, is delayed after the Retrier is evicted from the retriers map.
+                        // Not doing so will allow the user to manually retry before the Retrier has been deleted, making that retry to instantly fail.
+                        // Also, there is nothing to do here for Subscription errors.
+                        _ => (),
```

Comment:
> Not doing so will allow the user to manually retry before the Retrier has been deleted

It's now based on retrier being there in retriers map and not solely on tower status. Let's update this piece to:
`No need to set the tower status to unreachable because we already set it when idling above`
Or we can actually move the piece that sets the tower status to unreachable from the idle branch in `e.is_permanent()` check to this place instead.

Reply:
Umm, this made me realize that for `RetryError::Subscription(_, false)` we were setting the tower status to `Unreachable` and right after to `SubscriptionError` (given `RetryError::Subscription(_, false)` is not permanent).

So we'll need to change the match to `RetryError::Subscription(_, false)` and put the aforementioned code in the default branch:

```rust
log::warn!("Retry strategy gave up for {}. {}", self.tower_id, e);
if e.is_permanent() {
    self.set_status(RetrierStatus::Failed);
}

match e {
    RetryError::Subscription(_, false) => {
        log::info!("Setting {} status as subscription error", self.tower_id);
        self.wt_client
            .lock()
            .unwrap()
            .set_tower_status(self.tower_id, TowerStatus::SubscriptionError)
    }
    RetryError::Misbehaving(p) => {
        log::warn!("Cannot recover known tower_id from the appointment receipt. Flagging tower as misbehaving");
        self.wt_client
            .lock()
            .unwrap()
            .flag_misbehaving_tower(self.tower_id, p);
    }
    RetryError::Abandoned => {
        log::info!("Skipping retrying abandoned tower {}", self.tower_id)
    }
    // This covers `RetryError::Unreachable` and `RetryError::Subscription(_, false)`
    _ => {
        log::debug!("Starting to idle");
        self.set_status(RetrierStatus::Idle(Instant::now()));
        // Clear all pending appointments so they do not waste any memory while idling
        self.pending_appointments.lock().unwrap().clear();
        self.wt_client
            .lock()
            .unwrap()
            .set_tower_status(self.tower_id, TowerStatus::Unreachable);
    }
}
```

Reply:
Good point,

> So we'll need to change the match to RetryError::Subscription(_, false) and put the

I guess you meant `RetryError::Subscription(_, true)` and in the code snippet as well.

Reply:
indeed